README.md
<h1 align="center">PilotJS</h1>

<div align="center">

[![GitHub license](https://img.shields.io/github/license/Wave-Play/pilot?style=flat)](https://github.com/Wave-Play/pilot/blob/main/LICENSE) [![Tests](https://github.com/Wave-Play/pilot/workflows/CI/badge.svg)](https://github.com/Wave-Play/pilot/actions) ![npm](https://img.shields.io/npm/v/@waveplay/pilot) [![minizipped size](https://badgen.net/bundlephobia/minzip/@waveplay/pilot)](https://bundlephobia.com/result?p=@waveplay/pilot)

**NextJS for Expo & React Native**

Customizable, fast, and lightweight drop-in support for **[NextJS](https://nextjs.org/)** on native platforms

</div>

## Features

- Supports most **NextJS** features on native (powered by **NextJS** on web)
- Hooks for loading events with ability to override routes
- Optional placeholders during load events
- Can be used to control individual areas rather than entire screen

## Install

Using NPM

```bash
npm install @waveplay/pilot
```

Using Yarn

```bash
yarn add @waveplay/pilot
```

## Quick start

Update your `App.js` entry component to render `PilotArea`.

> `App.js`
```jsx
import { PilotArea } from '@waveplay/pilot';

const App = () => {
  // ... your code

  return (
    <PilotArea/>
  );
};
export default App;
```

Use the **PilotJS** CLI to build your routes.

```bash
npx pilotjs-cli build
```

You're now ready to use **PilotJS**!

## Basic usage

You can use **PilotJS** the same way as Next Router, except it now works on React Native and Expo projects!

```ts
// const router = useRouter();
const pilot = usePilot();

// router.push('/dashboard');
pilot.fly('/dashboard'); // or pilot.push('/dashboard');

// router.back();
pilot.back();

// router.reload();
pilot.reload();

// router.asPath;
pilot.getPath();

// router.query;
pilot.getQuery();
```

## Supported NextJS features

| Feature              | Support             | Description |
|----------------------|---------------------|-------------|
| `/_app`               | <center>✅</center> | Optional app wrapper is applied on pages when the `_app` route is registered. |
| `/404`               | <center>✅</center> | Rendered when page route cannot be found or when `notFound` is returned as `true` while loading props. |
| `/500`               | <center>✅</center> | Rendered when an error is thrown while loading props. |
| `/pages`             | <center>✅</center> | Automatically finds routes from `/pages` or `/src/pages` directory via `pilotjs-cli build` command. |
| `getServerSideProps` | <center>✅</center> | Calls this function and delegates props if it exists when loading new route. |
| `getStaticProps`     | <center>✅</center> | Calls this function and delegates props if it exists when loading new route. May be skipped if props were cached. |
| `revalidate`         | <center>✅</center> | Uses this field returned by your `getStaticProps` function to skip future loading for the time specified. |
| `i18n`               | <center>✅</center> | Supports `defaultLocale` and `locales` from next.config.js `i18n` fields. `locale` is passed as a variable in your getProps functions. See pilot-i18next package for full locale support. |
| `context`            | <center>🕒</center> | `context` is passed to `getServerSideProps` and `getStaticProps` functions.<br/><br/>Fields not supported yet:<br/>`preview`, `previewData`, `req`, `res` |
| `<Link>`             | <center>🕒</center> | Supports most functionality that `next/link` provides.<br/><br/>Props not supported yet:<br/>`passHref`, `prefetch`, `replace`, `scroll`, `shallow` |
| `redirects`             | <center>🕒</center> | Redirects are supported by returning a `redirect` object.<br/><br/>Redirects defined in `next.config.js` are not supported yet. |
| `useRouter`          | <center>🕒</center> | Supports most functionality using the `usePilot()` hook.<br/><br/>Fields not supported yet:<br/>`basePath`, `domainLocales`, `isFallback`, `isReady`, `isPreview`.<br/><br/>Functions not supported yet:<br/>`beforePopState()`, `events()`, `prefetch()`, `replace()` |
| `/api`     | <center>❌</center> | Not supported yet. |
| `getStaticPaths`     | <center>❌</center> | Not supported yet. |
| `getInitialProps`    | <center>❌</center> | Not supported yet. |
| `middleware`           | <center>❌</center> | Not supported yet. |
| `rewrites`           | <center>❌</center> | Not supported yet. |
| `withRouter`         | <center>❌</center> | Not supported yet. |

If a feature is not in the table above, that likely means there are no plans to support it or we simply missed it. 

Native apps work much differently than a server which means some features may not be possible to support. Please open an issue if you'd like to see a feature added.

## Credits

This project was originally developed as an internal router for [WavePlay](https://waveplay.com).

## Documentation

- [Advanced setup for NextJS](https://github.com/Wave-Play/pilot/blob/main/docs/advanced-nextjs.md)

## License

The MIT License.


CHANGELOG.md
# [2.0.0-beta.3](https://github.com/Wave-Play/pilot/compare/v2.0.0-beta.2...v2.0.0-beta.3) (2022-10-17)


### Bug Fixes

* **ci:** include devDependencies in release ([0c0d307](https://github.com/Wave-Play/pilot/commit/0c0d30714623b93d8d1e645543dc79f357ce65ab))



# [2.0.0-beta.2](https://github.com/Wave-Play/pilot/compare/v2.0.0-beta.1...v2.0.0-beta.2) (2022-10-17)


### Bug Fixes

* **ci:** re-enabled semantic-release plugins ([9aba282](https://github.com/Wave-Play/pilot/commit/9aba2827ea325c08b39240c27eedc4b3102aca44))



# [2.0.0-beta.1](https://github.com/Wave-Play/pilot/compare/v1.1.1...v2.0.0-beta.1) (2022-10-17)


### Bug Fixes

* **ci:** build semantic-release in next, alpha, and beta branches ([a3cc669](https://github.com/Wave-Play/pilot/commit/a3cc669dd33fdb2a0d05e37d83e63c19ed6aece3))
* **ci:** defined all branches to run semantic-release on ([3c6d444](https://github.com/Wave-Play/pilot/commit/3c6d4446d4fc468b7e5772a5d674a81bf7e8f6d7))
* **ci:** drop "next" from release branches ([e336e7f](https://github.com/Wave-Play/pilot/commit/e336e7f7aa005c287c78a8fe28fbc4f5a00dcfae))
* **ci:** include beta and alpha branches in semantic-release ([70f2c30](https://github.com/Wave-Play/pilot/commit/70f2c3057b5375fd25cd927b11924e90cf42371e))
* **ci:** remove "release" from package.json ([0896aa7](https://github.com/Wave-Play/pilot/commit/0896aa73905c1d3abc4b7e86ba66b0fd8dffdfd7))
* **ci:** removed semantic-release/git options ([e0f3e26](https://github.com/Wave-Play/pilot/commit/e0f3e26126a805bbcf9e712c6482e951721ed2d5))
* **ci:** renamed main branch to master ([fb19542](https://github.com/Wave-Play/pilot/commit/fb195425ef8af88b030c8665f52bcf4dab3bdb5c))
* **ci:** temporarily removed semantic-release plugins ([23fbfd3](https://github.com/Wave-Play/pilot/commit/23fbfd3056ade6f8b5483f96f7dbbeead843620b))
* **ci:** use semantic-release/npm plugin ([37d55e0](https://github.com/Wave-Play/pilot/commit/37d55e0145865ffab5885bdd0a3cddc01d8450ca))


### Code Refactoring

* **api:** removed deprecated PilotArea props & refactored default ([3a4d0ac](https://github.com/Wave-Play/pilot/commit/3a4d0ac5ee723739d9530604a5f21edd1e2a5023))


### Features

* **api:** now possible to bring own logger ([66b12d2](https://github.com/Wave-Play/pilot/commit/66b12d2173dcf8a3f3d5b6f8e8ef55fd9e1766a2))
* **cli:** New CLI & build process ([f892fef](https://github.com/Wave-Play/pilot/commit/f892fef84617749707268d42e7c93a23ff519a7f))
* **link:** New Link component ([86c5939](https://github.com/Wave-Play/pilot/commit/86c593951959048c921367744ab1cf32c508c79b))
* push() alias ([4400422](https://github.com/Wave-Play/pilot/commit/4400422c8e69e43db40fa13a795048e6c1dbe6bd))


### BREAKING CHANGES

* **api:** `logLevel` config field was replaced by `logger`.
* **api:** Deprecated PilotArea props have been removed as well as `default` prop from PilotRoute. Use `defaultPath` prop in PilotArea instead.



## [1.1.1](https://github.com/Wave-Play/pilot/compare/v1.1.0...v1.1.1) (2022-10-14)


### Bug Fixes

* **api:** allow next/router to handle url object ([88bdf2d](https://github.com/Wave-Play/pilot/commit/88bdf2d6164807c23ad1f3f57b2507987c6a0ba6))
* **area:** use correct pilot instance for named areas ([f2ba1fa](https://github.com/Wave-Play/pilot/commit/f2ba1fa3f7dcad4eb0a0536c96581c8d2afd9a5d))



# [1.1.0](https://github.com/Wave-Play/pilot/compare/v1.0.2...v1.1.0) (2022-09-22)


### Features

* **pkg:** 1.1.0 release ([ca7dad9](https://github.com/Wave-Play/pilot/commit/ca7dad9ad69e49cb7de098dedd940c5d021566e7))



## [1.0.2](https://github.com/Wave-Play/pilot/compare/v1.0.1...v1.0.2) (2022-09-18)


### Bug Fixes

* **pkg:** Build before running semantic-release ([b818288](https://github.com/Wave-Play/pilot/commit/b8182886158b1d705485d00b333d6851b9f021fd))
* **pkg:** fixed published build issues ([f900fba](https://github.com/Wave-Play/pilot/commit/f900fba350d7e4a818b177e0feab147b3aec0f7b))



# [1.0.0](https://github.com/Wave-Play/pilot/compare/8eab9108081e2f2751a377e567e0033762d6d878...v1.0.0) (2022-09-15)


### Bug Fixes

* **pkg:** Semantic release ([179e1ab](https://github.com/Wave-Play/pilot/commit/179e1ab2c6641c629c67536182a4c25338991381))
* **pkg:** Semantic release ([8eab910](https://github.com/Wave-Play/pilot/commit/8eab9108081e2f2751a377e567e0033762d6d878))





src/pilot.ts
/**
 * © 2022 WavePlay <dev@waveplay.com>
 */
import { NextRouter } from 'next/router';
import { ComponentProps, ComponentType, createElement, ReactElement } from 'react';
import { PilotRouteOptions } from './route';
import { RadixRouter } from './radix-router';
import { lru, LRU } from 'tiny-lru';
import { DataMap, Default404, Default500, generateNumber, Url } from './_internal';
import pino from 'pino';

interface ActionResult {
	page?: PilotPage
	redirect?: string
}

interface FlightOptions {
	action: (path: string) => Promise<ActionResult | null>
	addToStack?: boolean
}

export interface PilotConfig {
	id?: string
	cacheSize?: number
	i18n?: {
		defaultLocale: string
		locales: string[]
	}
	logLevel?: 'trace' | 'debug' | 'info' | 'warn' | 'error'
	nextRouter?: NextRouter | null
	router?: PilotRouter
}

export interface PilotFlyOptions {
	locale?: string | false
	scroll?: boolean
	shallow?: boolean
}

export interface PilotEvent {
	error?: any
	page?: PilotPage
	type: PilotEventType
}

export type PilotEventType = '*' | 'error' | 'load-complete' | 'load-start' | 'redirect'

export interface PilotHook {
	callback: (path: string, event: PilotEvent) => string | void
	id: number
	type: PilotEventType
}

export interface PilotPage {
	component: ComponentType
	params: DataMap
	props: ComponentProps<any>
	query: DataMap
}

export interface PilotRouter {
	addRoute: (route: PilotRouteOptions) => void
	find: (path: string, options: PilotRouterOptions) => PilotRouteResult
	removeRoute: (path: string) => void
	stats: () => {
		key: string
		numRoutes: number
		routes: PilotRouteOptions[]
	}
}

export interface PilotRouterOptions {
	pilot: Pilot
}

export interface PilotRouteResult extends PilotRouteOptions {
	params?: DataMap
	query?: DataMap
}

export interface PilotStateProps {
	visible?: boolean
}

export class Pilot {
	// Constructor parameters
	private _config: PilotConfig;

	// Loaded props cache
	private _cache: LRU<any>;

	// Handy lil' logger for debugging
	private _logger = pino();

	// Trackers
	private _currentLocale?: string;
	private _currentPage?: PilotPage;
	private readonly _hooks: PilotHook[] = [];
	private readonly _stack: string[] = [];

	constructor(config?: PilotConfig) {
		this._config = config || {};

		// Update log level based on preference
		this._logger.level = config?.logLevel || 'warn';

		// Create cache
		this._cache = lru(config?.cacheSize || 100);

		// Use built-in default router if none is specified
		if (!config?.router) {
			this._config.router = new RadixRouter();
		}

		// Assign default path if router was provided
		if (config?.nextRouter) {
			this._stack.push(config?.nextRouter.asPath);
		}

		this._logger.debug(`[${this._getId()}] New instance created`);
	}

	public addHook(event: PilotEventType, callback: (path: string, event: PilotEvent) => void): number {
		this._logger.debug(`[${this._getId()}] addHook()`);
		const id = generateNumber();
		this._hooks.push({
			callback,	id, type: event
		});
		return id;
	}

	public addRoute(route: PilotRouteOptions) {
		this._logger.debug(`[${this._getId()}] addRoute(${JSON.stringify(route)})`);
		if (!this._config.router) throw new Error('');
		this._config.router.addRoute(route);
	}

	public async back() {
		this._logger.debug(`[${this._getId()}] back()`);

		// Get previous page in stack
		this._stack.pop();
		const path = this._stack[this._stack.length - 1];

		return this._fly(path, {
			action: async (path: string) => {
				// Delegate to NextJS router if one exists; use internal otherwise
				if (this._config.nextRouter) {
					const event = eventWaiter('popstate');
					this._config.nextRouter.back();
					window.removeEventListener('popstate', await event);
					return null;
				} else {
					return await this._load(path);
				}
			},
			addToStack: false
		});
	}

	public config(config: PilotConfig) {
		this._logger.debug(`[${this._getId()}] config(${JSON.stringify(config)})`);

		// Update only defined properties from new config
		// You can still remove/reset a property by setting it to null
		for (const key in config) {
			if (config[key] !== undefined) {
				this._config[key] = config[key];
			}
		}

		// Update config for cache and logger separately
		this._cache.max = config?.cacheSize !== undefined ? config.cacheSize : 100;
		this._logger.level = config?.logLevel || 'warn';
		this._currentLocale = config?.i18n?.defaultLocale;
	}

	public getDefaultLocale(): string | undefined {
		return this._config.i18n?.defaultLocale;
	}

	public getLocale(): string | undefined {
		this._logger.debug(`[${this._getId()}] getLocale()`);
		return this._currentLocale;
	}

	public getLocales(): string[] | undefined {
		this._logger.debug(`[${this._getId()}] getLocales()`);
		return this._config.i18n?.locales;
	}

	public getParams(): DataMap {
		this._logger.debug(`[${this._getId()}] getParams()`);
		return this._currentPage?.params || {};
	}

	public getPath(options?: { includeLocale?: boolean }): string {
		this._logger.debug(`[${this._getId()}] getPath()`);
		const { includeLocale } = options || {};

		if (this._config.nextRouter) {
			return this._config.nextRouter.asPath;
		} else if (includeLocale && this._currentLocale) {
			return '/' + this._currentLocale + this._stack[this._stack.length - 1];
		} else {
			return this._stack[this._stack.length - 1];
		}
	}

	public getProps(): any {
		this._logger.debug(`[${this._getId()}] getProps()`);
		return this._currentPage?.props || {};
	}

	public getQuery(): DataMap {
		this._logger.debug(`[${this._getId()}] getQuery()`);
		if (this._config.nextRouter) {
			return this._config.nextRouter.query as DataMap;
		} else {
			return this._currentPage?.query || {};
		}
	}

	/**
	 * Equivalent to Next's router.push().
	 * Updates content with registered path component, also loading their props.
	 * If a placeholder has been specified, it'll render that while the props are fetching.
	 * 
	 * @param path 
	 */
	public async fly(url: Url, as?: string, options?: PilotFlyOptions) {
		this._logger.debug(`[${this._getId()}] fly(${JSON.stringify(url)})`);

		return this._fly(url, {
			action: async (path: string) => {
				// Delegate to NextJS router if one exists; use internal router otherwise
				if (this._config.nextRouter) {
					// Account for updated path (e.g. hook overrides)
					if (typeof url !== 'string') {
						url.pathname = path;
					}

					await this._config.nextRouter.push(url, as, options);
					return null;
				} else {
					return await this._load(path, options);
				}
			}
		});
	}

	public log(level: 'trace' | 'debug' | 'info' | 'warn' | 'error', message?: string, ...args: any[]) {
		this._logger[level](`[${this._getId()}] ` + message, ...args);
	}

	/**
	 * Alias for fly().
	 * 
	 * See {@link fly} for more details.
	 */
	public async push(url: Url, as?: string, options?: PilotFlyOptions) {
		return this.fly(url, as, options);
	}

	public async reload() {
		this._logger.debug(`[${this._getId()}] refresh()`);

		// Get previous page in stack
		const path = this._stack[this._stack.length - 1];

		return this._fly(path, {
			action: async (path: string) => {
				// Delegate to NextJS router if one exists; use internal otherwise
				if (this._config.nextRouter) {
					const event = eventWaiter('load');
					this._config.nextRouter.reload();
					window.removeEventListener('load', await event);
					return null;
				} else {
					return await this._load(path);
				}
			},
			addToStack: false
		});
	}

	public removeHook(id: number) {
		this._logger.debug(`[${this._getId()}] removeHook(${id})`);
		const index = this._hooks.findIndex(hook => hook.id === id);
		this._hooks.splice(index, 1);
	}

	public removeRoute(path: string) {
		this._logger.debug(`[${this._getId()}] removeRoute(${path})`);
		this._config.router?.removeRoute(path);
	}

	public render(): ReactElement | null {
		this._logger.debug(`[${this._getId()}] render()`);

		// Can't render anything if no page has been set
		if (!this._currentPage) return null;

		// Render whatever is currently set, plus the props!
		const { component, props } = this._currentPage;
		return createElement(component, props);
	}

	public stats() {
		return {
			id: this._getId(),
			logLevel: this._logger.level,
			nextRouter: !!this._config.nextRouter,
			path: this.getPath() || null,
			params: this.getParams(),
			routerStats: this._config.router?.stats(),
			query: this.getQuery(),
			stackLength: this._stack.length,
		}
	}

	/**
	 * Boilerplate function mainly used to de-duplicate a lot of logic.
	 * To be called by certain other functions like back(), fly(), and refresh().
	 * 
	 * @param path Path to navigate to. Mainly used for notifying listeners.
	 * @param options Should contain action to execute + other options.
	 */
	private async _fly(url: Url, options: FlightOptions) {
		const { action, addToStack = true } = options;

		// Get path from URL or use directly if string
		let path = typeof url === 'string' ? url : url.pathname;

		// Query values are important too, ya know
		if (typeof url !== 'string' && url.query) {
			let token = '?';
			for (let key in url.query) {
				path += `${token}${key}=${url.query[key]}`;
				token = '&';
			}
		}

		// Notify listeners and allow them to modify the path
		const originalPath = path;
		path = await this._notify(path, { type: 'load-start' });
		if (path !== originalPath) {
			this._logger.info(`[${this._getId()}] A hook has modified this path: ${path}`);
		}

		try {
			// Execute specified action
			const result = await action(path);

			// If redirected, recursively call this function again with the new path after notifying hooks
			if (result?.redirect) {
				this._notify(path, { type: 'redirect' });
				return await this._fly(result.redirect, options);
			}
			this._currentPage = result?.page;
		} catch (e) {
			// In case of error, exit early & notify listeners
			this._logger.error(`[${this._getId()}] Error while executing action:`, e);
			this._notify(path, {
				type: 'error',
				page: undefined
			});
		}

		// Add to stack
		if (addToStack) {
			this._stack.push(path);
			this._logger.debug(`[${this._getId()}] New stack size: ${this._stack.length}`);
		}

		// Notify listeners
		this._notify(path, {
			type: this._config.nextRouter || this._currentPage ? 'load-complete' : 'error',
			page: this._currentPage
		});
	}

	/**
	 * Convenience function to get the unique ID for this pilot.
	 */
	private _getId(): string {
		return `pilot${this._config.id ? '-' + this._config.id : ''}`;
	}

	/**
	 * Loads a page and its props by calling it's `getServerSideProps()` or `getStaticProps()` function.
	 * The result is then stored into _currentPage to be rendered via the `render()` function.
	 * 
	 * This will not create an instance of the page component, but rather just call the props function.
	 * The component instance will be created when the page is rendered.
	 * 
	 * @param path Path to load.
	 * @returns Component and props for the specified path. If no component is found, returns null.
	 */
	private async _load(path: string, options?: PilotFlyOptions): Promise<ActionResult> {
		const hasQuery = path.includes('?');

		// Handle locale prefix only as long as caller doesn't opt out
		if (options?.locale !== false) {
			// If this path starts with another registered locale, make sure to update the current locale
			const locale = options?.locale || this._config.i18n?.locales?.find(locale => 
				path.startsWith(`/${locale}/`)
				|| path === `/${locale}`
				|| (hasQuery && path.substring(0, path.indexOf('?')) === `/${locale}`)
			);
			if (locale && locale !== this._currentLocale) {
				this._logger.debug(`[${this._getId()}] Locale changed from ${this._currentLocale} to ${locale}`);
				this._currentLocale = locale;
			}

			// Paths are handled internally, so we need to strip the locale prefix or router won't match
			if (path.startsWith(`/${this._currentLocale}/`)) {
				path = path.replace(`/${this._currentLocale}`, '');
			} else if (path === `/${this._currentLocale}`) {
				path = '/';
			} else if (hasQuery && path.substring(0, path.indexOf('?')) === `/${this._currentLocale}`) {
				path = '/';
			}
		}

		// Look up the route data for this path
		let route = this._config.router?.find(path, { pilot: this });

		// If no route is found, load 404 page instead
		if (route) {
			this._logger.debug(`[${this._getId()}] Route found for path: ${path} ...`, route);
		} else if (path === '/404' || path === '/500') {
			// Assign a default 404 page if none is found for convenience and to avoid infinite loops
			this._logger.warn(`[${this._getId()}] Using default ${path} ...`);
			route = {
				component: path === '/404' ? Default404 : Default500,
				path, params: {}, query: {}
			};
		} else {
			this._logger.warn(`[${this._getId()}] No route found for path: ${path}`);
			return await this._load('/404');
		}

		// Call this page's internal server side or static props function
		// This will emulate a NextJS page load, but without the usual server parameters, so be aware
		let props: any;
		try {
			props = await this._loadProps(path, route);
		} catch (e) {
			this._logger.error(`[${this._getId()}] Error loading props for path: ${path}`, e);
			this._notify(path, { error: e, type: 'error' });
			return await this._load('/500');
		}

		if (props?.props) {
			this._logger.info(`[${this._getId()}] Loaded props for ${path}:`, Object.keys(props.props));
		} else if (props?.notFound) {
			this._logger.info(`[${this._getId()}] Route responded with "notFound": ${path}`);
			return await this._load('/404');
		} else if (props?.redirect?.destination) {
			this._logger.info(`[${this._getId()}] Route responded with "redirect": ${path} -> ${props.redirect.destination}`);
			return {
				redirect: props.redirect.destination
			};
		}

		return {
			page: {
				component: route.component,
				params: route.params || {},
				props: props?.props,
				query: route.query || {}
			}
		};
	}

	/**
	 * Calls route's getProps() function to load props for the specified path.
	 * This basically emulates NextJS' getServerSideProps() or getStaticProps() function.
	 * If a "revalidate" value is returned, it will be stored in our LRU memory cache.
	 * 
	 * @param path Path to load.
	 * @param route Route for the specified path.
	 * @returns Props for the specified path.
	 */
	private async _loadProps(path: string, route: PilotRouteResult): Promise<any> {
		let props: any = null;

		// Return early if there's nothing to load
		if (!route.getProps) {
			return props;
		}

		// See if we can find a cached version of this page's props
		const cacheKey = (this._currentLocale || '') + path + JSON.stringify(route.query);
		const cachedProps = this._cache.get(cacheKey);
		const isExpired = !cachedProps || cachedProps?.__pilot?.expires < Date.now();
		if (cachedProps && isExpired) {
			this._logger.debug(`[${this._getId()}] Cached props for ${path} have expired, removing from cache...`);
			this._cache.delete(cacheKey);
		} else if (cachedProps && !isExpired) {
			this._logger.debug(`[${this._getId()}] Found cached props for path: ${path}`);
			return cachedProps;
		}
		
		// Get props from route's getProps() function
		props = await route.getProps({
			locale: this._currentLocale,
			params: route.params || {},
			query: route.query || {},
			req: {} as any,
			res: {} as any,
			resolvedUrl: path
		});

		// If a "revalidate" value is returned, cache the props for that amount of time
		// This emulates ISR (Incremental Static Regeneration) from NextJS
		if (props?.revalidate) {
			this._logger.debug(`[${this._getId()}] Caching props for path: ${path} (revalidate: ${props.revalidate})`);
			this._cache.set(cacheKey, {
				...props,
				__pilot: {
					expires: Date.now() + props.revalidate * 1000
				}
			});
		}

		return props;
	}

	/**
	 * Notifies all added hooks of a certain event.
	 * This allows hooks to modify the path, or do other things in response to events.
	 * 
	 * @param path Path to notify listeners about.
	 * @param event Event to notify listeners about.
	 * @returns The modified path. If no hooks modify the path, this will be the same as the original.
	 */
	private async _notify(path: string, event: PilotEvent): Promise<string> {
		this._logger.debug(`[${this._getId()}] _notify(${path})`);
		for (let hook of this._hooks) {
			if (hook.type === event.type || hook.type === '*') {
				const result = hook.callback(path, event);
				if (result) {
					path = result;
				}
			}
		}

		return path;
	}
}

/**
 * Smol utility function meant to help us wait for a window event to fire.
 * This is super useful for waiting for window reload and back events!
 */
const eventWaiter = async (event: string): Promise<any> => {
	return new Promise<any>((resolve) => {
		const callback = () => resolve(callback);
		window.addEventListener(event, callback);
	});
};
